crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+//! despite checkpoints themselves potentially being executed nonsequentially and in parallel.
+//! CheckpointExecutor parallelizes checkpoints of the same epoch as much as possible, and
+//! handles epoch boundaries by calling to reconfig once all checkpoints of an epoch have finished
+//! executing.
+
+use core::panic;
+use std::{sync::Arc, time::Duration};
+
+use futures::stream::FuturesOrdered;
+use mysten_metrics::spawn_monitored_task;
+use sui_types::{
+    base_types::{ExecutionDigests, TransactionDigest},
+    crypto::AuthorityPublicKeyBytes,
+    error::{SuiError, SuiResult},
+    messages::TransactionEffects,
+    messages_checkpoint::{CheckpointSequenceNumber, VerifiedCheckpoint},
+};
+use tokio::{
+    sync::broadcast::{self, error::RecvError},
+    task::JoinHandle,
+};
+use tokio_stream::StreamExt;
+use tracing::{error, info, warn};
+use typed_store::rocks::TypedStoreError;
+
+use crate::{
+    authority::{AuthorityState, EffectsNotifyRead},
+    checkpoints::{CheckpointMetrics, CheckpointStore},
+};
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+const TASKS_PER_CORE: usize = 1;
+
+type CheckpointExecutionBuffer = FuturesOrdered<
+    JoinHandle<(
+        u64,
+        VerifiedCheckpoint,
+        Option<Vec<(AuthorityPublicKeyBytes, u64)>>,
+    )>,
+>;
+
+pub struct CheckpointExecutor {
+    mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+    checkpoint_store: Arc<CheckpointStore>,
+    authority_state: Arc<AuthorityState>,
+    highest_scheduled_checkpoint: Option<CheckpointSequenceNumber>,
+    end_of_epoch: bool,
+    metrics: Arc<CheckpointMetrics>,
+}
+
+impl CheckpointExecutor {
+    pub fn new(
+        mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+        metrics: Arc<CheckpointMetrics>,
+    ) -> Result<Self, TypedStoreError> {
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            authority_state,
+            highest_scheduled_checkpoint: None,
+            end_of_epoch: false,
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        while let Some((last_checkpoint, next_committee)) =
+            self.execute_checkpoints_for_epoch().await
+        {
+            reconfig(next_committee).await;
+            self.checkpoint_store
+                .update_highest_executed_checkpoint(&last_checkpoint)
+                .unwrap();
+            self.end_of_epoch = false;
+        }
+        // Channel closed
+    }
+
+    /// Executes all checkpoints for the current epoch. At epoch boundary,
+    /// awaits the queue of scheduled checkpoints, and returns the committe
+    /// of the next epoch.
+    pub async fn execute_checkpoints_for_epoch(
+        &mut self,
+    ) -> Option<(VerifiedCheckpoint, Vec<(AuthorityPublicKeyBytes, u64)>)> {
+        let mut finished: CheckpointExecutionBuffer = FuturesOrdered::new();
```

Comment:
Should this be called `pending` instead of `finished`? It stores tasks that are currently executing, right?